ingestor-default/src/main/scala/za/co/absa/hyperdrive/ingestor/implementation/transformer/deduplicate/kafka/DeduplicateKafkaSinkTransformer.scala

```diff
@@ -0,0 +1,200 @@
+import org.apache.spark.sql.functions.{col, lit, not, struct}
+import za.co.absa.hyperdrive.ingestor.api.transformer.{StreamTransformer, StreamTransformerFactory}
+import za.co.absa.hyperdrive.ingestor.api.utils.ConfigUtils.{getOrThrow, getPropertySubset, getSeqOrThrow}
+import za.co.absa.hyperdrive.ingestor.api.utils.StreamWriterUtil
+import za.co.absa.hyperdrive.ingestor.api.writer.StreamWriterCommonAttributes
+import za.co.absa.hyperdrive.ingestor.implementation.reader.kafka.KafkaStreamReader
+import za.co.absa.hyperdrive.ingestor.implementation.utils.AvroUtil
+import za.co.absa.hyperdrive.ingestor.implementation.writer.kafka.KafkaStreamWriter
+
+
+private[transformer] class DeduplicateKafkaSinkTransformer(
+  val readerTopic: String,
+  val readerBrokers: String,
+  val readerExtraOptions: Map[String, String],
+  val readerSchemaRegistryUrl: String,
+  val writerTopic: String,
+  val writerBrokers: String,
+  val writerExtraOptions: Map[String, String],
+  val writerSchemaRegistryUrl: String,
+  val checkpointLocation: String,
+  val sourceIdColumnNames: Seq[String],
+  val destinationIdColumnNames: Seq[String],
+  val kafkaConsumerTimeout: Duration
+) extends StreamTransformer {
+  private val logger = LogManager.getLogger
+
+  override def transform(dataFrame: DataFrame): DataFrame = {
+    val spark = dataFrame.sparkSession
+    val offsetLog = new OffsetSeqLog(spark, new Path(checkpointLocation, "offsets").toString)
+    val commitLog = new CommitLog(spark, new Path(checkpointLocation, "commits").toString)
+    val latestOffsetLog = offsetLog.getLatest().map(_._1)
+    val latestCommitLog = commitLog.getLatest().map(_._1)
+
+    if (latestOffsetLog != latestCommitLog) {
+      deduplicateDataFrame(dataFrame, offsetLog, commitLog)
+    } else {
+      dataFrame
+    }
+  }
+
+  private def deduplicateDataFrame(dataFrame: DataFrame, offsetLog: OffsetSeqLog, commitLog: CommitLog) = {
+    logger.info("Deduplicate rows after retry")
+    implicit val kafkaConsumerTimeoutImpl: Duration = kafkaConsumerTimeout
+    val sourceConsumer = createConsumer(readerBrokers, readerExtraOptions, readerSchemaRegistryUrl)
+    val latestCommittedOffsets = KafkaUtil.getLatestCommittedOffset(offsetLog, commitLog)
+    KafkaUtil.seekToOffsetsOrBeginning(sourceConsumer, readerTopic, latestCommittedOffsets)
+
+    val latestOffsetsOpt = KafkaUtil.getLatestOffset(offsetLog)
+    val sourceRecords = latestOffsetsOpt.map(latestOffset => consumeAndClose(sourceConsumer,
+      consumer => KafkaUtil.getMessagesAtLeastToOffset(consumer, latestOffset))).getOrElse(Seq())
+    val sourceIds = sourceRecords.map(extractIdFieldsFromRecord(_, sourceIdColumnNames))
+
+    val sinkConsumer = createConsumer(writerBrokers, writerExtraOptions, writerSchemaRegistryUrl)
+    val sinkTopicPartitions = KafkaUtil.getTopicPartitions(sinkConsumer, writerTopic)
+    val latestSinkRecords = consumeAndClose(sinkConsumer, consumer => sinkTopicPartitions.map {
```

Comment:
Can you do it for all partitions at once?

Reply:
Again, a very good question. Obviously, I thought no, but thinking again, it should be possible